src/poll_manager.ts

```diff
@@ -0,0 +1,149 @@
+    this.unsubscribeFunctions.add(this.subscribeVoteRemoved());
+  };
+
+  public unregisterSubscriptions = () => {
+    this.unsubscribeFunctions.forEach((cleanupFunction) => cleanupFunction());
+    this.unsubscribeFunctions.clear();
+  };
+
+  public createPoll = async (poll: CreatePollData<SCG>) => {
+    const { poll: createdPoll } = await this.client.createPoll(poll);
+
+    return new Poll({ client: this.client, poll: createdPoll });
+  };
+
+  public getPoll = async (id: string) => {
+    const { poll } = await this.client.getPoll(id);
+
+    this.setOrOverwriteInCache(poll, true);
+
+    return this.fromState(id);
+  };
+
+  public queryPolls = async (filter: QueryPollsFilters, sort: PollSort = [], options: QueryPollsOptions = {}) => {
+    const { polls, next } = await this.client.queryPolls(filter, sort, options);
+
+    const pollInstances = polls.map((poll) => {
+      this.setOrOverwriteInCache(poll, true);
+
+      return this.fromState(poll.id);
+    });
+
+    return {
+      polls: pollInstances,
+      next,
+    };
+  };
+
+  public hydratePollCache = (messages: FormatMessageResponse<SCG>[], overwriteState?: boolean) => {
+    for (const message of messages) {
+      if (!message.poll) {
+        continue;
+      }
+      const pollResponse = message.poll as PollResponse<SCG>;
+      this.setOrOverwriteInCache(pollResponse, overwriteState);
+    }
+  };
+
+  private setOrOverwriteInCache = (pollResponse: PollResponse<SCG>, overwriteState?: boolean) => {
+    const pollFromCache = this.fromState(pollResponse.id);
+    if (!pollFromCache) {
+      const poll = new Poll<SCG>({ client: this.client, poll: pollResponse });
+      this.pollCache.set(poll.id, poll);
+    } else if (overwriteState) {
+      pollFromCache.reinitializeState(pollResponse);
+    }
+  };
+
+  private subscribePollUpdated = () => {
+    return this.client.on('poll.updated', (event) => {
+      if (event.poll?.id) {
+        this.fromState(event.poll.id)?.handlePollUpdated(event);
+      }
+    }).unsubscribe;
+  };
+
+  private subscribePollClosed = () => {
+    return this.client.on('poll.closed', (event) => {
+      if (event.poll?.id) {
+        this.fromState(event.poll.id)?.handlePollClosed(event);
+      }
+    }).unsubscribe;
+  };
+
+  private subscribeVoteCasted = () => {
+    return this.client.on('poll.vote_casted', (event) => {
+      if (event.poll?.id) {
+        this.fromState(event.poll.id)?.handleVoteCasted(event);
+      }
+    }).unsubscribe;
+  };
+
+  private subscribeVoteChanged = () => {
+    return this.client.on('poll.vote_changed', (event) => {
+      if (event.poll?.id) {
+        this.fromState(event.poll.id)?.handleVoteChanged(event);
+      }
+    }).unsubscribe;
+  };
+
+  private subscribeVoteRemoved = () => {
+    return this.client.on('poll.vote_removed', (event) => {
+      if (event.poll?.id) {
+        this.fromState(event.poll.id)?.handleVoteRemoved(event);
+      }
+    }).unsubscribe;
+  };
```

Comment:
(including `subscribePollUpdated`)

Reply:
That's a no-go, it was indeed our initial approach however there are a few issues with it:

- Having `3000` polls essentially means having `3000 * 5` listeners at a certain point, all of which would be invoked when a new event arrives (and `2999` of them would simply skip because the event does not concern them)
- It doesn't particularly scale well with more polls being loaded through pagination
- This way, we only have a single set of listeners that consume one event at a time, updating the correct poll from within the cache directly (while maintaining referential integrity of the poll instances at all times)

So, in other words - they are indeed handled in the `Poll` instances themselves it's just not the `Poll` instances that listen to the events but rather the `manager` does.

Reply:
So esentially - poll instance cannot properly live without the manager - correct? You always have to rely on manager to keep your poll instance up to date so "detached" instances are "broken" in a way. 

I think it'd be good if we explored the idea of targeted events deeper - it should be very simple to do for polls as most of the events are poll-related and carry message/poll id's and we could benefit from this performance enhancement in the future. Poll could be fully encapsulated with logic it should "own" allowing it also to live in a detached (from the manager) state.

Reply:
If I understand you correctly, you're looking for something like:

- The `pollManager` is still listening for `poll` events
- It re-dispatches a targeted event (for example, `poll.vote_casted-<poll_id>` or something`)
- The poll itself is only listening to `poll.vote_casted-<this.id>` events and handles them

Is this correct ? I totally get your concerns and something like this was briefly considered too, however I do have some concerns about this approach too:

- Each `poll` having its own listener can potentially be a ton of extra listeners
- Not every listener would indeed be triggered (only the specific ones), but their event's `dispatch` would still be coupled with the `poll` manager, which would act as a `router` of sorts in order to orchestrate the traffic - doing these changes through events seems to me like a bit of an overhead considering that we can just update inline
- If `poll` instances are indeed to listen, they have to be subscribed to events; since we need all of these to be up to date at all times we would basically need to subscribe to the new events once again whenever `messages` actually arrive within the state `client.queryChannels` or `channel.query` for example)

With that said, I do see some really good things with this as well:

- The ability to hook up to specific `poll` events from anywhere (for example, if some custom integrator logic depends on events they can simply hook up to `poll.vote_casted-<id>` and not have to do an ID sanity check each time one of these events is fired)
- Having them detached is a very, very good point too

So, I'm not particularly sure which way we should go - perhaps we should sit down and discuss this more thoroughly. Adding targeted events shouldn't be difficult as far as I can see (pretty much the same logic).